From b18cca7036c7cb342a22611e78e1ba91019f91f3 Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Mon, 12 Dec 2022 16:38:55 +0100
Subject: [PATCH] imx8mm-var-dart: Integrate iMX8MM DART and SOM with balenaOS

This includes custom dtb support in u-boot
and is adapted to the updated u-boot 2021
provided with kernel 5.10+

Upstream-status: Inaproppriate [configuration]
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 configs/imx8mm_var_dart_defconfig |  9 +++++++++
 include/configs/imx8mm_var_dart.h | 32 +++++++++++++++++++++----------
 2 files changed, 31 insertions(+), 10 deletions(-)

diff --git a/configs/imx8mm_var_dart_defconfig b/configs/imx8mm_var_dart_defconfig
index 599eaf5c4e..22e4e0f11b 100644
--- a/configs/imx8mm_var_dart_defconfig
+++ b/configs/imx8mm_var_dart_defconfig
@@ -154,3 +154,12 @@ CONFIG_EFI_IGNORE_OSINDICATIONS=y
 CONFIG_EFI_CAPSULE_FIRMWARE_RAW=y
 CONFIG_EFI_SECURE_BOOT=y
 # CONFIG_OPTEE_LIB is not set
+CONFIG_CMD_IMPORTENV=y
+CONFIG_FS_UUID=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_SETEXPR=y
+CONFIG_CMD_EXT2=y
+CONFIG_CMD_EXT4=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_EXPORTENV=y
+CONFIG_PARTITION_UUIDS=y
diff --git a/include/configs/imx8mm_var_dart.h b/include/configs/imx8mm_var_dart.h
index 9368674a36..54fc7aefb6 100644
--- a/include/configs/imx8mm_var_dart.h
+++ b/include/configs/imx8mm_var_dart.h
@@ -109,28 +109,36 @@
 		"fi; " \
 		"bootaux ${m4_addr};\0" \
 	"optargs=setenv bootargs ${bootargs} ${kernelargs};\0" \
-	"mmcargs=setenv bootargs console=${console} " \
-		"root=/dev/mmcblk${mmcblk}p${mmcpart} rootwait rw ${cma_size} cma_name=linux,cma\0 " \
-	"loadbootscript=load mmc ${mmcdev}:${mmcpart} ${loadaddr} ${bootdir}/${bsp_script};\0" \
+	"mmcargs=setenv bootargs console=${console} ${resin_kernel_root} rootwait rw ${os_cmdline} " \
+		" ${cma_size}\0 " \
+	"loadbootscript=load mmc ${mmcdev}:${mmcpart} ${loadaddr} ${script};\0" \
 	"bootscript=echo Running bootscript from mmc ...; " \
 		"source\0" \
-	"loadimage=load mmc ${mmcdev}:${mmcpart} ${img_addr} ${bootdir}/${image};" \
+	"loadimage=load mmc ${mmcdev}:${resin_root_part} ${img_addr} boot/${image};" \
 		"unzip ${img_addr} ${loadaddr}\0" \
 	"findfdt=" \
 		"if test $fdt_file = undefined; then " \
-			"if test $board_name = VAR-SOM-MX8M-MINI; then " \
-				"setenv fdt_file imx8mm-var-som-symphony.dtb; " \
+			"if test -n ${custom_fdt_file}; then " \
+				"setenv fdt_file ${custom_fdt_file}; " \
 			"else " \
-				"if test $carrier_rev = legacy; then " \
-					"setenv fdt_file imx8mm-var-dart-dt8mcustomboard-legacy.dtb; " \
+				"if test $board_name = VAR-SOM-MX8M-MINI; then " \
+					"if test $carrier_rev = legacy; then " \
+						"setenv fdt_file imx8mm-var-som-symphony-legacy.dtb; " \
+					"else " \
+						"setenv fdt_file imx8mm-var-som-symphony.dtb; " \
+					"fi; " \
 				"else " \
-					"setenv fdt_file imx8mm-var-dart-dt8mcustomboard.dtb; " \
+					"if test $carrier_rev = legacy; then " \
+						"setenv fdt_file imx8mm-var-dart-dt8mcustomboard-legacy.dtb; " \
+					"else " \
+						"setenv fdt_file imx8mm-var-dart-dt8mcustomboard.dtb; " \
+					"fi; " \
 				"fi; " \
 			"fi; " \
 		"fi; \0" \
 	"loadfdt=run findfdt; " \
 		"echo fdt_file=${fdt_file}; " \
-		"load mmc ${mmcdev}:${mmcpart} ${fdt_addr} ${bootdir}/${fdt_file}\0" \
+		"load mmc ${mmcdev}:${resin_root_part} ${fdt_addr} boot/${fdt_file}\0" \
 	"ramsize_check="\
 		"if test $sdram_size -le 512; then " \
 			"setenv cma_size cma=320M; " \
@@ -178,6 +186,10 @@
 		"fi;\0" \
 	"bsp_bootcmd=echo Running BSP bootcmd ...; " \
 	"run ramsize_check; " \
+        "setenv resin_kernel_load_addr ${loadaddr};" \
+        "run resin_set_kernel_root; run set_os_cmdline; gpio set 137; " \
+        "setenv mmcdev ${resin_dev_index};" \
+        "setenv mmcbootpart ${resin_boot_part};" \
 	"mmc dev ${mmcdev}; "\
 	"if mmc rescan; then " \
 		"if test ${use_m4} = yes && run loadm4bin; then " \
-- 
2.37.2

